src/Integration.Vsix/ErrorList/IssuesSnapshot.cs

```diff
@@ -270,23 +272,37 @@ public override bool TryCreateDetailsStringContent(int index, out string content
             return true;
         }
 
+        /// <summary>
+        /// Called when a new snapshot has been created or when an existing snapshot's version was changed.
+        /// The Error List will then try to map from a selected item in the previous snapshot to the
+        /// corresponding item in the new snapshot.
+        /// The Error List will raise two "selection changed" events - firstly to
+        /// "null", then to the corresponding issue in the new snapshot.
+        /// </summary>
+        /// <remarks>
+        /// 1. We can only do this if the snapshots represent the same analysis (i.e. if <see cref="AnalysisRunId"/> is the same).
+        /// 2. The new snapshot can be the same instance as the old snapshot: this will happen when Increment() is called.
+        /// 3. If the snapshots are different, it's possible that they don't have the same set of issues:
+        /// If <see cref="CreateUpdatedSnapshot()"/> was called, it will remove any non-navigable issues.
+        /// </remarks>
         public override int IndexOf(int currentIndex, ITableEntriesSnapshot newSnapshot)
         {
-            // Called when a new snapshot has been created and the Error List is trying to
-            // map from a selected item in the previous snapshot to the corresponding item
-            // in the new snapshot.
-            // We can only do this if the snapshots represent the same set of issues i.e.
-            // the AnalysisRunIds are the same.
-            // The Error List will still raise two "selection changed" events - firstly to 
-            // "null", then to the corresponding issue in the new snapshot.
             if (newSnapshot is IssuesSnapshot newIssuesSnapshot &&
                 newIssuesSnapshot.AnalysisRunId == AnalysisRunId &&
-                currentIndex >= 0 && currentIndex < issues.Count && // defensive - shouldn't happen unless VS passes an invalid index
-                !ShouldHideIssue(issues[currentIndex]) // don't map hidden issues: see #2351
-                )
+                currentIndex >= 0 && currentIndex < issues.Count) // defensive - shouldn't happen unless VS passes an invalid index
             {
-                return currentIndex;
+                var issueInOldSnapshot = issues[currentIndex];
+                var issueInNewSnapshot = newIssuesSnapshot.issues.ElementAtOrDefault(currentIndex);
+
+                // perf optimization: attempt to find the issue in the same index before doing a full search
```

Comment:
Strictly speaking setting `issueInNewSnapshot` in line 295 is part of this optimisation.
It might be better to co-locate all of the optimisation-related code i.e. move line 295 below the comment.